### ISSUE TITLE
UsbYubiKeyDevice hangs without returning

### ISSUE THREAD

**Issue by slowcar** [Yubico/yubikit-android#158]
I am currently implementing S/MIME with Yubikey and working on decryption using JCA/Bouncycastle.
When a PIN is entered i request a connection, retrieve the private key and cert and decrypt the content using CMSEnvelopedDataParser
This works in most cases, but sometimes the call does not return.

Here is a part of my code (Java)
```
new Thread(() ->
	usbYubiKeyDevice.requestConnection(SmartCardConnection.class, value -> {
		try {
			SmartCardConnection smartCardConnection = value.getValue();
			PivSession pivSession = new PivSession(smartCardConnection);
                        PivProvider pivProvider = new PivProvider(pivSession);
            		Security.insertProviderAt(pivProvider, 1);
            		KeyStore keyStore = KeyStore.getInstance("YKPiv");
           		keyStore.load(null);
            		String keystoreAlias = Slot.SIGNATURE.getStringAlias();
            		PrivateKey privateKey = (PrivateKey) keyStore.getKey(keystoreAlias, yubikeyPin);
```
The console shows
UsbDevice...ctionJNI  E  device is closed in native_control_request
UsbDevice...ctionJNI  D  close

Stepping through the code shows that the getKey call leads to UsbYubiKeyDevice.java line 133 (second line of the quoted code block), but does not return from there - no IOException is thrown.
```
executorService.submit(() -> {
	try (T connection = connectionManager.openConnection(connectionType)) {
		callback.invoke(Result.success(connection));
	} catch (IOException e) {
		callback.invoke(Result.failure(e));
	}
});
```
The reason seems to be that engineGetKey in PivKeyStoreSpi.java catches an IOException "Failed to send 15 bytes" in the catch (Exception e) block, which throws a new RuntimeException.
```
} catch (Exception e) {
	throw new RuntimeException(e);
}
```
However, the executorService block does not catch this exception and just silently dies.

**Comment by AdamVe** [comment]
Hi, thanks for the report! I wonder if you can reproduce the issue which causes the  IOException "Failed to send 15 bytes" or if that is random.

**Comment by slowcar** [comment]
I suspect that i am missing some kind of handling for the device, it seems to work when i restart the app, but only once.
Even if i disconnect the device and reconnect it the error occurs.

The documentation does not state that you have to close anything, except for the stopUsbDiscovery which i call, i am checking the AndroidDemo for code that does something like that, but did not find anything - i am not very experienced using Kotlin though.

Using the NfcYubiKeyDevice has the same error, but there the IOException is propagated properly and can be handled.
Basically it means i have two issues: One that the error is handled incorrectly with the usb connection, and the second that it occurs at all.

**Comment by AdamVe** [comment]
Thanks for explaining. From your console logs I see the device is closed - that means your application has released the USB device resource. 

Can you see in your log when [this code](https://github.com/Yubico/yubikit-android/blob/release/2.7.0/android/src/main/java/com/yubico/yubikit/android/transport/usb/UsbYubiKeyDevice.java#L159) is called (I assume that you use yubikit-android 2.7.0)?

**Comment by slowcar** [comment]
UsbYubikeyDevice close() is called when i navigate out of the fragment and the model/presenter is destroyed.
It is also called when i physically remove the key.

**Comment by AdamVe** [comment]
I will need some time to figure out why this could be happening - if you have some minimal code which I could run, that would help a lot.

Did you try the test suites for PIV and PIV JCA we have in the repository?

**Comment by slowcar** [comment]
I will take a look at the test cases to check the implementation. I have seen that both the demo app and the test cases work with a BlockingQueue, i will try to implement it like that as well and see how it works.

**Comment by slowcar** [comment]
I think my issue occurs due to multiple threads/repeat connections.
My implementation used a thread to import the key, then return to the UI to ask for the pin. This started a new thread, using requestConnection again.
I have resolved this by using a second BlockingQueue to handle the pin entry.
Now i just have to get this working using NFC with the present key/remove key pattern.

In any case, even though the error is no longer occurring with the better/different implementation, i think the exception handling should be improved to propagate the exception.

**Comment by slowcar** [comment]
I am unable to reproduce the behavior, i'll close it
